src/state/actions/dispatchArNSUpdate.ts

```diff
@@ -3,140 +3,214 @@ import {
 
-export function dispatchArNSUpdate({
-  emitter,
+export async function dispatchArNSUpdate({
   dispatch,
   walletAddress,
   arioProcessId,
   ao,
   antAo,
   aoNetworkSettings,
 }: {
-  emitter: ArNSEventEmitter;
   dispatch: Dispatch<ArNSAction>;
   walletAddress: AoAddress;
   arioProcessId: string;
   ao: AoClient;
   antAo: AoClient;
   aoNetworkSettings: typeof NETWORK_DEFAULTS.AO;
 }) {
-  dispatch({ type: 'setDomains', payload: {} });
-  dispatch({ type: 'setAnts', payload: {} });
-  dispatch({ type: 'setPercentLoaded', payload: 0 });
-  dispatch({ type: 'setAntCount', payload: 0 });
-  dispatch({
-    type: 'setLoading',
-    payload: true,
-  });
-  emitter.on('process', async (id, process) => {
-    const handlers = await queryClient.fetchQuery({
-      queryKey: ['handlers', id, aoNetworkSettings.ANT],
-      queryFn: async () => {
-        // validate transfer supports eth addresses and single char undername fix and max ttl
-        const drySetRecordRes = await antAo
-          .dryrun({
-            process: id,
-            Owner: walletAddress.toString(),
-            From: walletAddress.toString(),
-            tags: [
-              { name: 'Action', value: 'Set-Record' },
-              { name: 'Sub-Domain', value: 'a' },
-              { name: 'Transaction-Id', value: 'test-'.padEnd(43, '1') },
-              { name: 'TTL-Seconds', value: '86400' },
-            ],
-          })
-          .catch(() => {
-            return {} as ReturnType<typeof ao.dryrun>;
-          });
+  try {
+    const limit = pLimit(20);
+    // reset queries
+    queryClient.resetQueries({
+      queryKey: ['ant'],
+    });
+    queryClient.resetQueries({
+      queryKey: ['handlers'],
+    });
+    dispatch({ type: 'reset' });
+    dispatch({ type: 'setDomains', payload: {} });
+    dispatch({ type: 'setAnts', payload: {} });
+    dispatch({ type: 'setPercentLoaded', payload: 0 });
+    dispatch({ type: 'setAntCount', payload: 0 });
+    dispatch({
+      type: 'setLoading',
+      payload: true,
+    });
+    const arioContract = ARIO.init({
+      process: new AOProcess({ processId: arioProcessId, ao }),
+    }) as AoARIORead;
+    const antRegistry = ANTRegistry.init({
+      process: new AOProcess({
+        processId: ANT_REGISTRY_ID,
+        ao: connect(aoNetworkSettings.ANT),
+      }),
+    });
+    const [arnsRecords, allUserAnts] = await Promise.all([
+      arioContract.getArNSRecords({ limit: 100_000 }),
+      antRegistry.accessControlList({ address: walletAddress.toString() }),
+    ]);
 
-        const hasError =
-          drySetRecordRes.Messages.find((msg) => {
-            return msg.Tags.find((t: Tag) => t.name === 'Error');
-          }) !== undefined;
+    const flatAntIds = new Set([
+      ...allUserAnts.Controlled,
+      ...allUserAnts.Owned,
+    ]);
 
-        if (hasError) {
-          return [] as AoANTHandler[];
+    const userDomains = arnsRecords.items.reduce(
```

Comment:
In general the scheme to download all records then filter isn't going to scale. We need a different path moving forward, perhaps adding something to the SDK/contract to allow querying passed-in list of ids or something of that sort.

Reply:
agree we'll want something like this - but will punt it to post launch